### PR TITLE
[Bugfix] Fix flaky test by storing feedbacks in the correct way

### PR DESCRIPTION
**Pull request by stefanwaldhauser**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context

`de.tum.in.www1.artemis.programmingexercise.ProgrammingExerciseResultJenkinsIntegrationTest.shouldGenerateNewManualResultIfManualAssessmentExists(ProgrammingExerciseResultJenkinsIntegrationTest.java:89)`

The problem is explained in detail here: https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/database/?highlight=%40OrderColumn#best-practices. When saving an ordered collection, you basically have to first save the collection entities first on their own before saving the whole collection.

### Description

Fixed the problem by storing the feedback in the correct way

### Steps for Testing

Go through the whole workflow of an programming exercise and check that everything works (create pe, take part in ce and get an automatic result, change the automatic result in the manual assessment)

### Test Coverage

- no changes
### Screenshots

- no changes

### CODE REVIEW

**Comment by krusche** [comment]
Some server tests fail now, so apparently the change was not without consequences, please have another look

**Comment by stefanwaldhauser** [comment]
> Some server tests fail now, so apparently the change was not without consequences, please have another look

Very strange. I will look into it!

**Comment by stefanwaldhauser** [comment]
> Some server tests fail now, so apparently the change was not without consequences, please have another look

9dbf70e02ce5fb5d0dc99da390221f4f206aff36 
So the error was basically that in the old implementation the new feedbacks were added and in mine I replaced all the feeedacks. This should fix it now and restore the old behaviour

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 1]
```diff
@@ -188,16 +189,46 @@ public Result createNewExampleResultForSubmissionWithExampleSubmission(long subm
...
+     * <p>
+     * With ordered collections (like result and feedback here), we have to be very careful with the way we persist the objects in the database.
+     * We must first persist the child object without a relation to the parent object. Then, we recreate the association and persist the parent object.
      *
+     * If the result is not saved (shouldSave = false), the caller is responsible to save the result (which will persist the feedback changes as well)
+     *
+     * @param result           the result with should be saved with the given feedback
+     * @param feedbackList     new feedback items which replace the existing feedback
+     * @param shouldSave       whether the result should be saved or not
+     * @return the update (and potentially saved) result
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java:202]:
```suggestion
     * @return the updated (and potentially saved) result
```

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 2]
```diff
@@ -188,16 +189,46 @@ public Result createNewExampleResultForSubmissionWithExampleSubmission(long subm
...
-     * @param feedbackList new feedback items which should be saved to the feedback
-     * @param shouldSaveResult whether the result should be saved or not
-     * @return the saved result
+     *
+     * If the result is not saved (shouldSave = false), the caller is responsible to save the result (which will persist the feedback changes as well)
+     *
+     * @param result           the result with should be saved with the given feedback
+     * @param feedbackList     new feedback items which should be added to the feedback
+     * @param shouldSave       whether the result should be saved or not
+     * @return the update (and potentially saved) result
```
Comment (FuchsDominik) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java:221]:
```suggestion
     * @return the updated (and potentially saved) result
```

**Review by FuchsDominik — DISMISSED** [review]
Code looks good

**Review by FrankeLukas — DISMISSED** [review]
Changes lgtm.

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java [hunk 3]
```diff
@@ -165,10 +172,8 @@ private Result updateLatestSemiAutomaticResultWithNewAutomaticFeedback(long last
         latestSemiAutomaticResult.getFeedbacks().removeIf(feedback -> feedback != null && feedback.getType() == FeedbackType.AUTOMATIC);
 
         // copy all feedback from the automatic result
-        for (Feedback feedback : newAutomaticResult.getFeedbacks()) {
-            Feedback newFeedback = feedback.copyFeedback();
-            latestSemiAutomaticResult.addFeedback(newFeedback);
-        }
+        List<Feedback> copiedFeedbacks = newAutomaticResult.getFeedbacks().stream().map(Feedback::copyFeedback).collect(Collectors.toList());
```
Comment (joschahenningsen) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseGradingService.java:175]: nice!

**Review by joschahenningsen — APPROVED** [review]
looks good :+1:

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 4]
```diff
@@ -188,16 +189,46 @@ public Result createNewExampleResultForSubmissionWithExampleSubmission(long subm
...
+     * We must first persist the child object without a relation to the parent object. Then, we recreate the association and persist the parent object.
      *
+     * If the result is not saved (shouldSave = false), the caller is responsible to save the result (which will persist the feedback changes as well)
+     *
+     * @param result           the result with should be saved with the given feedback
+     * @param feedbackList     new feedback items which replace the existing feedback
+     * @param shouldSave       whether the result should be saved or not
+     * @return the update (and potentially saved) result
+     */
+    public Result storeFeedbackInResult(@NotNull Result result, List<Feedback> feedbackList, boolean shouldSave) {
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java:204]:
```suggestion
    public Result storeFeedbacksInResult(@NotNull Result result, List<Feedback> feedbackList, boolean shouldSave) {
```

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 5]
```diff
@@ -188,16 +189,46 @@ public Result createNewExampleResultForSubmissionWithExampleSubmission(long subm
...
+     * If the result is not saved (shouldSave = false), the caller is responsible to save the result (which will persist the feedback changes as well)
+     *
+     * @param result           the result with should be saved with the given feedback
+     * @param feedbackList     new feedback items which should be added to the feedback
+     * @param shouldSave       whether the result should be saved or not
+     * @return the update (and potentially saved) result
      */
-    public Result storeFeedbackInResult(Result result, List<Feedback> feedbackList, boolean shouldSaveResult) {
+    @NotNull
+    public Result addFeedbackToResult(@NotNull Result result, List<Feedback> feedbackList, boolean shouldSave) {
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java:224]:
```suggestion
    public Result addFeedbacksToResult(@NotNull Result result, List<Feedback> feedbackList, boolean shouldSave) {
```

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java [hunk 6]
```diff
@@ -188,16 +189,46 @@ public Result createNewExampleResultForSubmissionWithExampleSubmission(long subm
...
+        List<Feedback> savedFeedbacks = saveFeedbackWithHibernateWorkaround(result, feedbackList);
+        result.addFeedbacks(savedFeedbacks);
+        return shouldSaveResult(result, shouldSave);
+    }
+
+    @NotNull
+    private List<Feedback> saveFeedbackWithHibernateWorkaround(@NotNull Result result, List<Feedback> feedbackList) {
```
Comment (ivanchimeno) [inline src/main/java/de/tum/in/www1/artemis/service/ResultService.java:231]:
```suggestion
    private List<Feedback> saveFeedbacksWithHibernateWorkaround(@NotNull Result result, List<Feedback> feedbackList) {
```

**Review by ivanchimeno — CHANGES_REQUESTED** [review]
Looks also good to me but I added minor suggesstions to rename the functions :)

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java (hunk 4 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ResultService.java:204]: Feedback does not really have a plural, so I keep the method name

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java (hunk 5 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ResultService.java:224]: Feedback does not really have a plural, so I keep the method name

File: src/main/java/de/tum/in/www1/artemis/service/ResultService.java (hunk 6 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/ResultService.java:231]: Feedback does not really have a plural, so I keep the method name

**Review by krusche — APPROVED** [review]

**Pull request merged**